sui_programmability/framework/sources/Governance/Delegation.move

```diff
@@ -0,0 +1,139 @@
+    use Sui::Coin::{Self, Coin};
+    use Sui::ID::{Self, VersionedID};
+    use Sui::SUI::SUI;
+    use Sui::Transfer;
+    use Sui::TxContext::{Self, TxContext};
+
+    friend Sui::SuiSystem;
+
+    /// A custodial delegation object. When the delegation is active, the delegation
+    /// object holds the delegated stake coin. It also contains the delegation
+    /// target validator address.
+    /// The delegation object is required to claim delegation reward. The object
+    /// keeps track of the next reward unclaimed epoch. One can only claim reward
+    /// for the epoch that matches the `next_reward_unclaimed_epoch`.
+    /// When the delegation is deactivated, we keep track of the ending epoch
+    /// so that we know the ending epoch that the delegator can still claim reward.
+    struct Delegation has key {
+        id: VersionedID,
+        /// The delegated stake, if the delegate is still active
+        active_delegation: Option<Coin<SUI>>,
+        /// If the delegation is inactive, `ending_epoch` will be
+        /// set to the ending epoch, i.e. the epoch when the delegation
+        /// was withdrawn. Delegator will not be eligible to claim reward
+        /// for ending_epoch and after.
+        ending_epoch: Option<u64>,
+        /// The delegated stake amount.
+        delegate_amount: u64,
+        /// Delegator is able to claim reward epoch by epoch. `next_reward_unclaimed_epoch`
+        /// is the next epoch that the delegator can claim epoch. Whenever the delegator
+        /// claims reward for an epoch, this value increments by one.
+        next_reward_unclaimed_epoch: u64,
+        /// The delegation target validator.
+        validator_address: address,
+    }
+
+    public(friend) fun create(
+        starting_epoch: u64,
+        validator_address: address,
+        stake: Coin<SUI>,
+        ctx: &mut TxContext,
+    ) {
+        let delegate_amount = Coin::value(&stake);
+        let delegation = Delegation {
+            id: TxContext::new_id(ctx),
+            active_delegation: Option::some(stake),
+            ending_epoch: Option::none(),
+            delegate_amount,
+            next_reward_unclaimed_epoch: starting_epoch,
+            validator_address,
+        };
+        Transfer::transfer(delegation, TxContext::sender(ctx))
+    }
+
+    /// Deactivate the delegation. Send back the stake and set the ending epoch.
+    public(friend) fun undelegate(
+        self: &mut Delegation,
+        ending_epoch: u64,
+        ctx: &mut TxContext,
+    ) {
+        assert!(is_active(self), 0);
+
+        let stake = Option::extract(&mut self.active_delegation);
+        let sender = TxContext::sender(ctx);
+        Transfer::transfer(stake, sender);
+
+        self.ending_epoch = Option::some(ending_epoch);
+    }
+
+    /// Checks whether the delegation object is eligible to claim the reward
+    /// given the epoch to claim and the validator address.
+    public(friend) fun can_claim_reward(
```

Comment:
Does this function need to be a `friend` if it doesn't mutate state? It's not important, but still. Maybe someone could write some helper/wrapper for delegation?